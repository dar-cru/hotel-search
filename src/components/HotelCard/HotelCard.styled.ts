import styled from 'styled-components';

export const FlexContainer = styled.div<{ direction: 'row' | 'column' }>`
  display: flex;
  flex-direction: ${(props) => props.direction || 'row'};
  position: relative;
  box-sizing: border-box;
`;

export const Divider = styled.hr`
  margin: 0;
  background-color: #cbcaca;
  border: 0;
  height: 1px;
`;

export const TitleHeaderContainer = styled.div`
  display: flex;
  flex-direction: row;
  width: 70%;
  padding-top: 20px;
`;

export const HeadingText = styled.h2`
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  margin: 0;
  padding-right: 8px;
`;

export const GreyInfoText = styled.p`
  color: #999999;
  margin: 0;
  font-size: 0.8rem;
`;

export const RoomOfferText = styled.p`
  position: absolute;
  bottom: 18px;
  text-decoration: underline;
  color: #e40000;
  font-size: 0.8rem;
  margin: 0;
`;

export const CurrencyText = styled.p`
  font-size: 2rem;
  color: #3f3f3f;
  font-weight: 200;
  padding: 4px 0;
  margin: 0;

  :before {
    content: '$';
    font-size: 1.15rem;
    vertical-align: 60%;
    padding-right: 2px;
  }
`;

export const SavingsText = styled.p`
  font-size: 1rem;
  color: #e40000;
  margin: 0;

  :after {
    content: '~';
    font-size: 0.8rem;
    vertical-align: 25%;
    padding-left: 1px;
  }
`;

export const CancellationTypeText = styled.p`
  position: absolute;
  margin: 0;
  bottom: 8px;
  color: #35a509;
  font-size: 0.8rem;
`;
