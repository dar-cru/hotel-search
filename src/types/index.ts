export * from './hotel';
export * from './offer';
export * from './property';
export * from './rating';
