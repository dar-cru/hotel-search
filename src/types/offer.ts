export enum PromotionType {
  CAMPAIGN = 'CAMPAIGN',
  MEMBER = 'MEMBER'
}

export enum CancellationType {
  NOT_REFUNDABLE = 'NOT_REFUNDABLE',
  FREE_CANCELLATION = 'FREE_CANCELLATION'
}

export type Offer = {
  promotion: {
    title: string;
    type: PromotionType;
  };
  name: string;
  displayPrice: {
    amount: number;
    currency: string;
  };
  savings?: {
    amount: number;
    currency: string;
  };
  cancellationOption: {
    cancellationType: CancellationType;
  };
};
