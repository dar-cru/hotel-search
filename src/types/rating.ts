export enum RatingType {
  SELF = 'self',
  STAR = 'star'
}

export type Rating = {
  ratingValue: 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5;
  ratingType: RatingType;
};
