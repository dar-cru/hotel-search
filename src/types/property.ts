import { Rating } from './rating';

export type Property = {
  propertyId: string;
  title: string;
  address: string[];
  previewImage: {
    url: string;
    caption: string;
    imageType: string;
  };
  rating: Rating;
};
