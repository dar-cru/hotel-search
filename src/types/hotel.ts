import { Offer } from './offer';
import { Property } from './property';

export type Hotel = {
  id: string;
  property: Property;
  offer: Offer;
};
