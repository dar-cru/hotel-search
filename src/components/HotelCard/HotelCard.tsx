import { CancellationType, Offer, Property } from '../../types';
import ImageTextOverlap from '../ImageTextOverlap';
import Rating from '../Rating/Rating';
import {
  Divider,
  HeadingText,
  GreyInfoText,
  CurrencyText,
  SavingsText,
  TitleHeaderContainer,
  FlexContainer,
  RoomOfferText,
  CancellationTypeText
} from './HotelCard.styled';

type HotelCardProps = {
  property: Property;
  offer: Offer;
};

const HotelCard = ({ property, offer }: HotelCardProps) => {
  const { title, address, previewImage, rating } = property;
  const { promotion, name: offerName, cancellationOption, displayPrice, savings } = offer;
  return (
    <FlexContainer direction="row" style={{ paddingBottom: '5px' }}>
      <ImageTextOverlap url={previewImage.url} imgDesc={previewImage.caption} text={promotion.title} />

      <div style={{ width: '100%', padding: '0 10px' }}>
        <Divider />

        <TitleHeaderContainer>
          <HeadingText>{title}</HeadingText>
          <Rating {...rating} />
        </TitleHeaderContainer>

        <FlexContainer direction="row" style={{ minHeight: '116px' }}>
          <FlexContainer direction="column" style={{ maxHeight: '70px', paddingTop: '4px', width: '50%' }}>
            <GreyInfoText>{address.join(', ')}</GreyInfoText>
            <RoomOfferText>{offerName}</RoomOfferText>
          </FlexContainer>

          <FlexContainer direction="column" style={{ alignItems: 'flex-end', paddingTop: '28px', width: '50%' }}>
            <GreyInfoText>1 night total ({displayPrice.currency?.toUpperCase()})</GreyInfoText>
            <CurrencyText>{displayPrice.amount}</CurrencyText>
            {savings && <SavingsText>Save ${savings.amount}</SavingsText>}
          </FlexContainer>
        </FlexContainer>

        {cancellationOption?.cancellationType === CancellationType.FREE_CANCELLATION && (
          <CancellationTypeText>Free Cancellation</CancellationTypeText>
        )}
      </div>
    </FlexContainer>
  );
};

export default HotelCard;
