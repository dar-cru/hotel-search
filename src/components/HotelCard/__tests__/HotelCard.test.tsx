import { render, RenderResult, screen } from '@testing-library/react';
import { CancellationType, Offer, PromotionType, Property, RatingType } from '../../../types';
import HotelCard from '../HotelCard';

const mockHotelProperty: Property = {
  propertyId: 'P107802',
  title: 'Primus Hotel Sydney',
  address: ['339 Pitt St', 'Sydney'],
  previewImage: {
    url: 'https://mockUrl.com',
    caption: 'Image of Primus Hotel Sydney',
    imageType: 'PRIMARY'
  },
  rating: {
    ratingValue: 5,
    ratingType: RatingType.SELF
  }
};

const mockHotelOffer: Offer = {
  promotion: {
    title: 'Exclusive Deal',
    type: PromotionType.MEMBER
  },
  name: 'Deluxe King',
  displayPrice: {
    // eslint-disable-next-line prettier/prettier
    amount: 375.000000,
    currency: 'AUD'
  },
  savings: {
    amount: 28.0,
    currency: 'AUD'
  },
  cancellationOption: {
    cancellationType: CancellationType.FREE_CANCELLATION
  }
};

describe('Hotel Card component', () => {
  it('renders with the correct values through props', () => {
    const tree: RenderResult = render(<HotelCard property={mockHotelProperty} offer={mockHotelOffer} />);

    const { container } = tree;
    const hotelAddress = screen.queryByTestId('hotel-address-text');
    const nightTotalCurrency = screen.queryByTestId('hotel-night-total-text');
    const pricePerNight = screen.queryByTestId('hotel-price-text');
    const hotelSavings = screen.queryByTestId('hotel-savings-text');
    const freeCancellation = screen.queryByTestId('hotel-free-cancellation');

    expect(container).toMatchSnapshot();
    expect(hotelAddress?.textContent).toEqual('339 Pitt St, Sydney');
    expect(pricePerNight?.textContent).toEqual('375');
    expect(nightTotalCurrency?.textContent).toEqual('1 night total (AUD)');
    expect(hotelSavings?.textContent).toEqual('Save $28');
    expect(freeCancellation?.textContent).toEqual('Free Cancellation');
  });

  it('should nicely format all of all address values', () => {
    render(
      <HotelCard
        property={{ ...mockHotelProperty, address: ['121 Pitt St', 'Sydney', 'NSW'] }}
        offer={mockHotelOffer}
      />
    );

    const hotelAddress = screen.queryByTestId('hotel-address-text');
    expect(hotelAddress?.textContent).toEqual('121 Pitt St, Sydney, NSW');
  });

  it('should not render the address if the values are empty', () => {
    render(<HotelCard property={{ ...mockHotelProperty, address: [] }} offer={mockHotelOffer} />);

    const hotelAddress = screen.queryByTestId('hotel-address-text');
    expect(hotelAddress?.textContent).toEqual('');
  });

  it('should render the correct currency and rounded off display price', () => {
    render(
      <HotelCard
        property={mockHotelProperty}
        offer={{
          ...mockHotelOffer,
          displayPrice: {
            currency: 'NZD',
            amount: 299.99
          }
        }}
      />
    );

    const nightTotalCurrency = screen.queryByTestId('hotel-night-total-text');
    const pricePerNight = screen.queryByTestId('hotel-price-text');

    expect(pricePerNight?.textContent).toEqual('300');
    expect(nightTotalCurrency?.textContent).toEqual('1 night total (NZD)');
  });

  it('should round the savings value correctly', () => {
    const { rerender } = render(
      <HotelCard
        property={mockHotelProperty}
        offer={{
          ...mockHotelOffer,
          savings: {
            amount: 99.29,
            currency: 'AUD'
          }
        }}
      />
    );

    let savingsPrice = screen.queryByTestId('hotel-savings-text');
    expect(savingsPrice).not.toBeNull();
    expect(savingsPrice?.textContent).toEqual('Save $99');

    rerender(
      <HotelCard
        property={mockHotelProperty}
        offer={{
          ...mockHotelOffer,
          savings: {
            amount: 99.79,
            currency: 'AUD'
          }
        }}
      />
    );

    savingsPrice = screen.queryByTestId('hotel-savings-text');
    expect(savingsPrice).not.toBeNull();
    expect(savingsPrice?.textContent).toEqual('Save $100');

    rerender(
      <HotelCard
        property={mockHotelProperty}
        offer={{
          ...mockHotelOffer,
          savings: {
            amount: 99.5111119,
            currency: 'AUD'
          }
        }}
      />
    );

    savingsPrice = screen.queryByTestId('hotel-savings-text');
    expect(savingsPrice).not.toBeNull();
    expect(savingsPrice?.textContent).toEqual('Save $100');
  });

  it('should not display the savings text if the value is null', () => {
    render(
      <HotelCard
        property={mockHotelProperty}
        offer={{
          ...mockHotelOffer,
          savings: undefined
        }}
      />
    );

    const savingsPrice = screen.queryByTestId('hotel-savings-text');
    expect(savingsPrice).toBeNull();
    expect(savingsPrice?.textContent).toBeUndefined();
  });

  it('should not render the free cancellation text if cancellation type is non-refundable', () => {
    render(
      <HotelCard
        property={mockHotelProperty}
        offer={{
          ...mockHotelOffer,
          cancellationOption: {
            cancellationType: CancellationType.NOT_REFUNDABLE
          }
        }}
      />
    );

    const freeCancellation = screen.queryByTestId('hotel-free-cancellation');
    expect(freeCancellation).toBeNull();
    expect(freeCancellation?.textContent).toBeUndefined();
  });
});
