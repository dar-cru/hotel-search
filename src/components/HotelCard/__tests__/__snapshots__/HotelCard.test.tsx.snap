// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hotel Card component renders with the correct values through props 1`] = `
.c1 {
  position: relative;
  width: 242px;
  height: 150px;
  margin: 5px 10px 5px 0;
}

.c2 {
  position: absolute;
  width: 100%;
  height: 100%;
}

.c3 {
  position: absolute;
  background-color: white;
  padding: 6px;
  top: 12px;
  left: 0;
  box-sizing: border-box;
  color: red;
  margin: 0;
  font-size: 0.9em;
  font-weight: 500;
}

.c7 {
  display: inline-block;
  font-size: 1.5rem;
  font-family: Times;
  line-height: 1;
}

.c7:before {
  content: '●●●●●';
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  background: linear-gradient(90deg,gold 100%,grey 100%);
  -webkit-background-clip: text;
  -webkit-text-fill-color: transparent;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  position: relative;
  box-sizing: border-box;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  box-sizing: border-box;
}

.c4 {
  margin: 0;
  background-color: #cbcaca;
  border: 0;
  height: 1px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  width: 70%;
  padding-top: 20px;
}

.c6 {
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  margin: 0;
  padding-right: 8px;
}

.c9 {
  color: #999999;
  margin: 0;
  font-size: 0.8rem;
}

.c10 {
  position: absolute;
  bottom: 18px;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #e40000;
  font-size: 0.8rem;
  margin: 0;
}

.c11 {
  font-size: 2rem;
  color: #3f3f3f;
  font-weight: 200;
  padding: 4px 0;
  margin: 0;
}

.c11:before {
  content: '$';
  font-size: 1.15rem;
  vertical-align: 60%;
  padding-right: 2px;
}

.c12 {
  font-size: 1rem;
  color: #e40000;
  margin: 0;
}

.c12:after {
  content: '~';
  font-size: 0.8rem;
  vertical-align: 25%;
  padding-left: 1px;
}

.c13 {
  position: absolute;
  margin: 0;
  bottom: 8px;
  color: #35a509;
  font-size: 0.8rem;
}

<div>
  <div
    class="c0"
    direction="row"
    style="padding-bottom: 5px;"
  >
    <div
      class="c1"
    >
      <img
        alt="Image of Primus Hotel Sydney"
        class="c2"
        src="https://mockUrl.com"
      />
      <p
        class="c3"
      >
        Exclusive Deal
      </p>
    </div>
    <div
      style="width: 100%; padding: 0px 10px;"
    >
      <hr
        class="c4"
      />
      <div
        class="c5"
      >
        <h2
          class="c6"
        >
          Primus Hotel Sydney
        </h2>
        <div
          class="c7"
        />
      </div>
      <div
        class="c0"
        direction="row"
        style="min-height: 116px;"
      >
        <div
          class="c8"
          direction="column"
          style="max-height: 70px; padding-top: 4px; width: 50%;"
        >
          <p
            class="c9"
            data-testid="hotel-address-text"
          >
            339 Pitt St, Sydney
          </p>
          <p
            class="c10"
          >
            Deluxe King
          </p>
        </div>
        <div
          class="c8"
          direction="column"
          style="align-items: flex-end; padding-top: 28px; width: 50%;"
        >
          <p
            class="c9"
            data-testid="hotel-night-total-text"
          >
            1 night total (
            AUD
            )
          </p>
          <p
            class="c11"
            data-testid="hotel-price-text"
          >
            375
          </p>
          <p
            class="c12"
            data-testid="hotel-savings-text"
          >
            Save $
            28
          </p>
        </div>
      </div>
      <p
        class="c13"
        data-testid="hotel-free-cancellation"
      >
        Free Cancellation
      </p>
    </div>
  </div>
</div>
`;
